File: problem/binarytrees/solution/go/8.go
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// Go adaptation of binary-trees Rust #4 program
// Use semaphores to match the number of workers with the CPU count
//
// contributed by Marcel Ibes
// modified by Isaac Gouy
// modified by Adam Shaver(use the struct constructor for bottomUpTree)
//
//

package main

import (
    "context"
    "flag"
    "fmt"
    "math"
    "runtime"
    "sort"
    "strconv"

    "golang.org/x/sync/semaphore"
)

type Tree struct {
    Left  *Tree
    Right *Tree
}

type Message struct {
    Pos  uint32
    Text string
}

var unity = uint32(1)

type ByPos []Message

func (m ByPos) Len() int           { return len(m) }
func (m ByPos) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
func (m ByPos) Less(i, j int) bool { return m[i].Pos < m[j].Pos }

func itemCheck(tree *Tree) uint32 {
    if tree.Left != nil && tree.Right != nil {
        return uint32(1) + itemCheck(tree.Right) + itemCheck(tree.Left)
    }

    return 1
}

func bottomUpTree(depth uint32) *Tree {
    if depth > uint32(0) {
        return &Tree{Left: bottomUpTree(depth - 1), Right: bottomUpTree(depth -
1)}
    } else {
        return &Tree{}
    }
}

func inner(depth, iterations uint32) string {
    chk := uint32(0)
    for i := uint32(0); i < iterations; i++ {
        a := bottomUpTree(depth)
        chk += itemCheck(a)
    }
    return fmt.Sprintf("%d\t trees of depth %d\t check: %d",
        iterations, depth, chk)
}

const minDepth = uint32(4)

func main() {
    n := 0
    flag.Parse()
    if flag.NArg() > 0 {
        n, _ = strconv.Atoi(flag.Arg(0))
    }

    run(uint32(n))
}

func run(n uint32) {
    cpuCount := runtime.NumCPU()
    sem := semaphore.NewWeighted(int64(cpuCount))

    maxDepth := n
    if minDepth+2 > n {
        maxDepth = minDepth + 2
    }

    depth := maxDepth + 1

    messages := make(chan Message, cpuCount)
    expected := uint32(2) // initialize with the 2 summary messages

    go func() {
        // do stretch tree and longLivedTree
        if err := sem.Acquire(context.TODO(), 1); err == nil {
            go func() {
                defer sem.Release(1)
                tree := bottomUpTree(depth)
                messages <- Message{0,
                    fmt.Sprintf("stretch tree of depth %d\t check: %d",
                        depth, itemCheck(tree))}
            }()
        } else {
            panic(err)
        }

        if err := sem.Acquire(context.TODO(), 1); err == nil {
            go func() {
                defer sem.Release(1)
                longLivedTree := bottomUpTree(maxDepth)
                messages <- Message{math.MaxUint32,
                    fmt.Sprintf("long lived tree of depth %d\t check: %d",
                        maxDepth, itemCheck(longLivedTree))}
            }()
        } else {
            panic(err)
        }

        for halfDepth := minDepth / 2; halfDepth < maxDepth/2+1; halfDepth++ {
            depth := halfDepth * 2
            iterations := uint32(1 << (maxDepth - depth + minDepth))
            expected++

            func(d, i uint32) {
                if err := sem.Acquire(context.TODO(), 1); err == nil {
                    go func() {
                        defer sem.Release(1)
                        messages <- Message{d, inner(d, i)}
                    }()
                } else {
                    panic(err)
                }
            }(depth, iterations)
        }
    }()

    var sortedMsg []Message
    for m := range messages {
        sortedMsg = append(sortedMsg, m)
        expected--
        if expected == 0 {
            close(messages)
        }
    }

    sort.Sort(ByPos(sortedMsg))
    for _, m := range sortedMsg {
        fmt.Println(m.Text)
    }
}

File: problem/binarytrees/solution/dartexe/8.dart
/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   transliterated from Andrey Filatkin's Node #6 program by Isaac Gouy
*/

import 'dart:isolate';

void main(List<String> args) {
    final maxDepth = (args.length > 0) ? int.parse(args[0]) : 6;

    final stretchDepth = maxDepth + 1;
    final check = itemCheck(bottomUpTree(stretchDepth));
    print("stretch tree of depth $stretchDepth\t check: $check");

    final longLivedTree = bottomUpTree(maxDepth);

    final requests = <Request>[];
    for (var depth = 4; depth <= maxDepth; depth += 2) {
        final iterations = 1 << maxDepth - depth + 4;
        requests.add(Request(iterations,depth));
    }
    var awaitedReplies = requests.length; // each request should receive a reply

    ReceivePort _replyPort = ReceivePort();
    var nIsolates = 4;
    while (nIsolates-- > 0) {
        Isolate.spawn(requestReply,_replyPort.sendPort);
    }

    final replies = <Reply>[]; // will not arrive in order, so cache & sort

    _replyPort.listen((dynamic message) {
        if (message is SendPort) {
            if (requests.length > 0) {
                message.send(requests.removeLast());
            }
        } else {
            replies.add(message as Reply);
            awaitedReplies--;

            if (awaitedReplies < 1){
                replies.sort((a,b) => a.depth.compareTo(b.depth));
                replies.forEach((r) =>
                    print("${r.iterations}\t trees of depth ${r.depth}\t check:
${r.check}")
                );
                print("long lived tree of depth $maxDepth\t check: ${itemCheck(l
ongLivedTree)}");

                _replyPort.close();
            }
        }
    });
}

void requestReply(SendPort p) {
    ReceivePort _requestPort = ReceivePort();
    p.send(_requestPort.sendPort);

    _requestPort.listen((dynamic message) {
        if (message is Request) {
            final iterations = message.iterations;
            final depth = message.depth;
            var check = 0;
            for (int i = 0; i < iterations; i++) {
                check += itemCheck(bottomUpTree(depth));
            }
            p.send(Reply(iterations,depth,check));
            p.send(_requestPort.sendPort);
        }
    });
}

class Request {
    int iterations = 0, depth = 0;
    Request(this.iterations,this.depth);
}

class Reply {
    int iterations = 0, depth = 0, check = 0;
    Reply(this.iterations,this.depth,this.check);
}

class TreeNode {
    TreeNode? left, right;
    TreeNode(this.left,this.right);
}

int itemCheck(TreeNode? node) {
    if (node == null || node.left == null) {
        return 1;
    } else {
        return 1 + itemCheck(node.left) + itemCheck(node.right);
    }
}

TreeNode bottomUpTree(int depth) {
    return depth > 0
        ? TreeNode(bottomUpTree(depth - 1), bottomUpTree(depth - 1))
        : TreeNode(null, null);
}

